Properly deprecate types.string

infinisil opened the pull request:
##### Motivation for this change

For almost 6 (!) years now, `types.string` has been marked as deprecated in the comment above its definition. People keep using this type regardless, and it's getting annoying while reviewing PRs to constantly have to remind people to not use it.

This PR does following things:
- Remove all usages of `types.string` in nixpkgs and replace them with a more appropriate type
- Make `types.string` emit a warning upon use

The mass replacement was done all manually, because I needed to decide what the replacement type should be. This was either
- `types.str` when merging doesn't make sense (e.g. `user` options)
- `types.separatedString " "` if its an option representing command line arguments (ideally it should be `listOf str` for new modules)
- `types.lines` if it's a line-based configuration format (where joining with newlines makes sense)
- `types.str` if it's a different configuration format where concatenation doesn't work

Ping @Profpatsch @rycee @aanderse @danbst 

###### Things done

I have **not** tested any of the modules. While there's a very low chance of any of these changes breaking anything, I might have made a mistake somewhere. A double-check would therefore be good. 

- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))

infinisil committed:
nixos/modules: Remove all usages of types.string

And replace them with a more appropriate type

Also fix up some minor module problems along the way

rycee commented:
Wonderful! I've recently been thinking about doing this exact thing since I see people accidentally using `types.string` in new modules all the time. It's a waste of energy to always have to look for and point this out when reviewing.

I'll make a proper review on the 12th 🙂

danbst commented on nixos/modules/programs/xss-lock.nix:
```diff
@@ -12,7 +12,7 @@ in
       default = "${pkgs.i3lock}/bin/i3lock";
       example = literalExample ''''${pkgs.i3lock-fancy}/bin/i3lock-fancy'';
-      type = types.string;
+      type = types.str;
```
maybe `types.path`? Can this even have cmd args here? @Mic92

Infinisil replied:
Yeah I think so, look at the usage of the option a bit later (is called with `-- ${lockerCommand}` in a CLI

Infinisil replied:
I guess it should be `separatedString " "` then

danbst commented on nixos/modules/services/databases/couchdb.nix:
```diff
@@ -138,7 +138,7 @@ in {
 
       configFile = mkOption {
-        type = types.string;
+        type = types.str;
```
`types.path`

Infinisil replied:
I'm not sure about switching to `types.path` for config files, because this has different behavior when the path is unquoted:
```nix
{
  configFile = /var/lib/foo.conf;
}
```

With `types.str` it will be set to `"/var/lib/foo.conf"`, a runtime path, whereas with `types.path` the file will be imported into the store and then that path is used, which can be unsafe if you have passwords stored in the file

danbst commented on nixos/modules/services/databases/openldap.nix:
```diff
@@ -47,26 +47,26 @@ in
 
       dataDir = mkOption {
-        type = types.string;
+        type = types.str;
```
```suggestion
        type = types.path;
```

Infinisil replied:
Same reason for maybe not using `path` here.

Tbh, maybe `types.path` should be an alias for `types.str`

Infinisil replied:
Ah your suggestion in https://github.com/NixOS/nixpkgs/pull/66346#discussion_r312350479 sounds better

danbst commented on nixos/modules/services/logging/SystemdJournal2Gelf.nix:
```diff
@@ -25,7 +25,7 @@ in
 
       extraOptions = mkOption {
-        type = types.string;
+        type = types.str;
```
```suggestion
        type = types.separatedString " ";
```
Just for consistency with postgresql-backup service, which you've changed to `separatedString`

danbst commented on nixos/modules/services/mail/exim.nix:
```diff
@@ -42,15 +42,15 @@ in
 
       spoolDir = mkOption {
-        type = types.string;
+        type = types.str;
```
```suggestion
        type = types.path;
```

danbst replied:
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

danbst commented on nixos/modules/services/misc/uhub.nix:
```diff
@@ -83,7 +83,7 @@ in
 	  };
           file = mkOption {
-            type = types.string;
+            type = types.str;
```
```suggestion
            type = types.path;
```

danbst commented on nixos/modules/services/misc/uhub.nix:
```diff
@@ -96,7 +96,7 @@ in
           };
           file = mkOption {
-            type = types.string;
+            type = types.str;
```
```suggestion
            type = types.path;
```

danbst commented on nixos/modules/services/networking/charybdis.nix:
```diff
@@ -21,30 +21,30 @@ in
 
       config = mkOption {
-        type = types.string;
+        type = types.str;
```
```suggestion
        type = types.lines;
```

danbst replied:
but I'm not sure the service supports multiple elements in config.

Infinisil replied:
Yeah it doesn't, I checked, see https://charybdis.readthedocs.io/en/release-3.5/config.html

danbst commented on nixos/modules/services/networking/charybdis.nix:
```diff
@@ -21,30 +21,30 @@ in
 
       statedir = mkOption {
-        type = types.string;
+        type = types.str;
```
I wanted to suggest `types.path`, but then I've recalled about a problem this may introduce - https://github.com/NixOS/nixpkgs/pull/57168.

So maybe some new type `types.strPath` or `types.statePath` should be introduced, which is `either str path` but does check if string starts with `/`, and does `toString path` if path object was provided.

danbst commented on nixos/modules/services/networking/gogoclient.nix:
```diff
@@ -34,7 +34,7 @@ in
       password = mkOption {
         default = "";
-        type = types.string;
+        type = types.str;
```
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

Infinisil replied:
Left this as `types.str` because the default is `""`

danbst commented on nixos/modules/services/networking/kippo.nix:
```diff
@@ -26,22 +26,22 @@ rec {
       varPath = mkOption {
         default = "/var/lib/kippo";
-        type = types.string;
+        type = types.str;
```
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

danbst commented on nixos/modules/services/networking/kippo.nix:
```diff
@@ -26,22 +26,22 @@ rec {
       logPath = mkOption {
         default = "/var/log/kippo";
-        type = types.string;
+        type = types.str;
```
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

danbst commented on nixos/modules/services/networking/kippo.nix:
```diff
@@ -26,22 +26,22 @@ rec {
       pidPath = mkOption {
         default = "/run/kippo";
-        type = types.string;
+        type = types.str;
```
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

danbst commented on nixos/modules/services/networking/prosody.nix:
```diff
@@ -297,7 +297,7 @@ in
 
       dataDir = mkOption {
-        type = types.string;
+        type = types.str;
```
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

danbst commented on nixos/modules/services/networking/radicale.nix:
```diff
@@ -41,7 +41,7 @@ in
 
     services.radicale.config = mkOption {
-      type = types.string;
+      type = types.str;
```
```suggestion
      type = types.lines;
```

Infinisil replied:
I thought about that, but concatenation doesn't have that great behavior for ini-like files (which radicale uses)

danbst replied:
yeah, just checked - it doesn't perform merge

danbst commented on nixos/modules/services/networking/softether.nix:
```diff
@@ -50,7 +50,7 @@ in
 
       dataDir = mkOption {
-        type = types.string;
+        type = types.str;
```
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

danbst commented on nixos/modules/services/security/munge.nix:
```diff
@@ -19,7 +19,7 @@ in
       password = mkOption {
         default = "/etc/munge/munge.key";
-        type = types.string;
+        type = types.str;
```
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

danbst commented on nixos/modules/services/security/oauth2_proxy.nix:
```diff
@@ -523,7 +523,7 @@ in
 
     keyFile = mkOption {
-      type = types.nullOr types.string;
+      type = types.nullOr types.str;
```
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

danbst commented on nixos/modules/services/web-apps/youtrack.nix:
```diff
@@ -73,7 +73,7 @@ in
         Where to keep the youtrack database.
       '';
-      type = types.string;
+      type = types.str;
```
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

danbst commented on nixos/modules/services/web-servers/uwsgi.nix:
```diff
@@ -72,7 +72,7 @@ in {
 
       runDir = mkOption {
-        type = types.string;
+        type = types.str;
```
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

danbst commented on nixos/modules/services/web-servers/zope2.nix:
```diff
@@ -23,19 +23,19 @@ let
       clientHome = mkOption {
         default = "/var/lib/zope2/${name}";
-        type = types.string;
+        type = types.str;
```
see https://github.com/NixOS/nixpkgs/pull/66346/files#r312350479

danbst commented on nixos/modules/services/web-servers/zope2.nix:
```diff
@@ -52,7 +52,7 @@ let
           </zodb_db>
           '';
-        type = types.string;
+        type = types.str;
```
it is kinda XML, but it is still extensible, accroding to https://github.com/kiorky/plone2/blob/master/etc/zope.conf.in
```suggestion
        type = types.lines;
```

danbst commented on nixos/modules/services/x11/hardware/synaptics.nix:
```diff
@@ -44,19 +44,19 @@ in {
 
       accelFactor = mkOption {
-        type = types.nullOr types.string;
+        type = types.nullOr types.str;
```
this is more of a joke, but...
```suggestion
        type = with types; nullOr (coercedTo float (toString) str);
```

Infinisil replied:
No that's a good suggestion, but I think I'll leave this out of scope for this PR

Profpatsch commented:
Great! I won’t have time for a review, but we should run all nixos tests to check for obvious problems. If a module has no test, we can’t be expected to do more than a best-effort thing.

Infinisil commented:
Hold on, I have this in my head that `types.path` is worse than `types.str` for sensitive files because it could accidentally put files in the world-readable /nix/store. But I can't actually reproduce this, `types.str` always seems to behave the same as `types.path` no matter what I do:

```nix
nix-repl> types.str.merge [] [ { value = ./default.nix; } ]
/home/infinisil/src/nixpkgs/default.nix

nix-repl> types.path.merge [] [ { value = ./default.nix; } ]
/home/infinisil/src/nixpkgs/default.nix

nix-repl> "${types.str.merge [] [ { value = ./default.nix; } ]}"
"/nix/store/9zpky4yjghjn02aprqfnz7ymfwvy156g-default.nix"

nix-repl> "${types.path.merge [] [ { value = ./default.nix; } ]}"
"/nix/store/9zpky4yjghjn02aprqfnz7ymfwvy156g-default.nix"

nix-repl> builtins.typeOf (types.str.merge [] [ { value = ./default.nix; } ])
"path"

nix-repl> builtins.typeOf (types.path.merge [] [ { value = ./default.nix; } ])
"path"
```

I feel a bit stupid now for having recommended `types.path` for sensitive things all this time. If I'm missing something I'd love to know

danbst commented:
@Infinisil the second to last is strange, because
```
nix-repl> builtins.typeOf "${/etc/nixos/configuration.nix}"
"string"
```

And yes, `types.path` is bad choice for password files. It was first reported in https://discourse.nixos.org/t/re-use-existing-home-migrating-from-ubuntu/2353

pacien commented:
`${/some/path}` copies `/some/path` to the store, whereas `${toString /some/path}` doesn't ([stackoverflow/43850371](https://stackoverflow.com/a/43850372)).

A safe option definition could look like:
```nix
{
  secretFile = mkOption {
    type = types.path;
    apply = toString;
  };
}
```
but a `types.strPath` or `types.externalPath` would be more explicit.

Infinisil commented:
Oh I did a mixup there in my last sentence, should've been

> I feel a bit stupid now for having recommended `types.str` for sensitive things all this time

Okay so then it seems that `types.str` vs `types.path` doesn't really matter for sensitive things, I'll apply the changes you suggested @danbst. The only solution for sensitive things would be to have modules use `toString` on it first. Although, this is also a bit of a problem, because it makes it non-trivial to add things to the store explicitly:

```nix
{
  # someOption uses `toString` on its value
  someOption = /foo/bar; # Not in store
  someOption = "/foo/bar"; # Not in store
  someOption = builtins.toPath /foo/bar; # Not in store
  someOption = builtins.toPath "/foo/bar"; # Not in store
  someOption = builtins.path { path = /foo/bar; }; # In store
  someOption = builtins.path { path = "/foo/bar"; }; # In store
}
```

And therefore is also not backwards compatible, so I'd rather not do that in this PR.

Infinisil commented:
Alright I switched all paths to the `types.path` type (unless the default was an empty string, in which case it's left as `types.str`).

Infinisil commented:
Whoops, did the commit history that you can't see the changes I made, here they are for reference: 224327dcbf0

rycee commented on nixos/modules/services/networking/smokeping.nix:
```diff
@@ -73,19 +73,19 @@ in
       cgiUrl = mkOption {
-        type = types.string;
+        type = types.str;
         default = "http://${cfg.hostName}:${builtins.toString cfg.port}/smokeping.cgi";
```
Hmm, this line seems problematic. It will cause documentation rebuilds if these values are in ones configuration.

CC @fadenb

rycee commented on nixos/modules/services/networking/smokeping.nix:
```diff
@@ -122,13 +122,13 @@ in
       imgUrl = mkOption {
-        type = types.string;
+        type = types.str;
         default = "http://${cfg.hostName}:${builtins.toString cfg.port}/cache";
```
Same as for `cgiUrl` above.

rycee commented on nixos/modules/services/x11/desktop-managers/surf-display.nix:
```diff
@@ -69,15 +69,15 @@ in {
 
       screensaverSettings = mkOption {
-        type = types.string;
+        type = types.str;
```
I think this could be `separatedString " "` since its simply passed on as [command line options to xset](https://code.it-zukunft-schule.de/cgit/surf-display/tree/bin/surf-display?id=972d6c4b7c4503dbb63fa6c92cdc24d1e32064a4#n182).

rycee reviewed (approved):
Really nice! I just added some minor comments.

Infinisil commented on nixos/modules/services/monitoring/dd-agent/dd-agent.nix:
```diff
@@ -145,31 +145,31 @@ in {
     postgresqlConfig = mkOption {
       description = "Datadog PostgreSQL integration configuration";
       default = null;
-      type = types.uniq (types.nullOr types.string);
```
I usually like to have `null` mean no value. While it doesn't make a difference here, for `nullOr path`, there wouldn't be a sensible null value of the type just `path`.

I'll change this to `nullOr lines` though

Infinisil commented on nixos/modules/programs/xss-lock.nix:
```diff
@@ -12,7 +12,7 @@ in
       default = "${pkgs.i3lock}/bin/i3lock";
       example = literalExample ''''${pkgs.i3lock-fancy}/bin/i3lock-fancy'';
-      type = types.string;
+      type = types.separatedString " ";
```
For most CLI's it does make sense to be able to pass multiple `--flags` and such and often they're not ordered either, so I think this makes sense to use

Infinisil commented:
@rycee Addressed all points. Also quickly fixed up the manual rebuild problem you mentioned in https://github.com/NixOS/nixpkgs/pull/66346#discussion_r313125997 and other couple minor things

infinisil committed:
lib/types: Make usage of types.string emit a warning

Infinisil commented:
And another new two instances of `types.string` being used.. in https://github.com/NixOS/nixpkgs/pull/67213 and in https://github.com/NixOS/nixpkgs/pull/66606

I fixed those up as well and rebased on master. Can we merge this?

@GrahamcOfBorg eval

rycee commented:
I looked through the latest changes and agree with them all. I'll merge if ofborg is happy.

RyzeNGrind commented:
Hi any updates on removing the types.string deprecation error? I am just now getting into Nix and NIxOS ecosystem and I cannot rebuild my configuration due to multiple module/nixpkg references coming from this deprecated type error. Kindly advise. 

```
[nixos@shogun:/mnt/c/Users/$USER/WSL]$ sudo nixos-rebuild switch --flake /etc/nixos#
building the system configuration...
error: The type `types.string` is deprecated. See https://github.com/NixOS/nixpkgs/pull/66346 for better alternative types.
(use '--show-trace' to show detailed location information)
```

wkordalski commented:
Even now [https://nixos.wiki/wiki/Declaration]() informs that `types.string` is the right thing to use.
I don't know if this site is yours or not, but for sure a significant number of NixOS beginners learns there how to use `mkOption`.
Putting correct information there might help with the repeating `types.string`.

Mic92 commented:
> https://nixos.wiki/wiki/Declaration

Everyone can edit their, including you. I fixed it now.

DaniD3v commented:
> types.separatedString " " if its an option representing command line arguments (ideally it should be listOf str for new modules)

Can you explain why `listOf str` should be used?

Should it be manually merged with `" "` in the `config = {}` section? Should we use `apply` in mkOptions?
Why is seperatedString even bad in the first place?

infinisil commented:
@DaniD3v `separatedString " "` breaks passing of arguments with spaces, because `concatStringsSep " " [ "foo bar" ] == concatStringsSep " " [ "foo" "bar" ]`. It should be `listOf str` so that you can use `lib.escapeShellArgs` to turn it into a string that bash properly interprets :)
